SignalMessaging/environment/OWSSounds.m

```diff
@@ -0,0 +1,463 @@
+    [instance.dbConnection setObject:@(sound)
+                              forKey:kOWSSoundsStorageGlobalNotificationKey
+                        inCollection:kOWSSoundsStorageNotificationCollection];
+}
+
++ (OWSSound)notificationSoundForThread:(TSThread *)thread
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    NSNumber *_Nullable value =
+        [instance.dbConnection objectForKey:thread.uniqueId inCollection:kOWSSoundsStorageNotificationCollection];
+    // Default to the "global" notification sound, which in turn will default to the global default.
+    return (value ? (OWSSound)value.intValue : [self globalNotificationSound]);
+}
+
++ (void)setNotificationSound:(OWSSound)sound forThread:(TSThread *)thread
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    [instance.dbConnection setObject:@(sound)
+                              forKey:thread.uniqueId
+                        inCollection:kOWSSoundsStorageNotificationCollection];
+}
+
+#pragma mark - Ringtones
+
++ (OWSSound)defaultRingtoneSound
+{
+    return OWSSound_Opening;
+}
+
++ (OWSSound)globalRingtoneSound
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    NSNumber *_Nullable value = [instance.dbConnection objectForKey:kOWSSoundsStorageGlobalRingtoneKey
+                                                       inCollection:kOWSSoundsStorageRingtoneCollection];
+    // Default to the global default.
+    return (value ? (OWSSound)value.intValue : [self defaultRingtoneSound]);
+}
+
++ (void)setGlobalRingtoneSound:(OWSSound)sound
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    [instance.dbConnection setObject:@(sound)
+                              forKey:kOWSSoundsStorageGlobalRingtoneKey
+                        inCollection:kOWSSoundsStorageRingtoneCollection];
+}
+
++ (OWSSound)ringtoneSoundForThread:(TSThread *)thread
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    NSNumber *_Nullable value =
+        [instance.dbConnection objectForKey:thread.uniqueId inCollection:kOWSSoundsStorageRingtoneCollection];
+    // Default to the "global" ringtone sound, which in turn will default to the global default.
+    return (value ? (OWSSound)value.intValue : [self globalRingtoneSound]);
+}
+
++ (void)setRingtoneSound:(OWSSound)sound forThread:(TSThread *)thread
+{
+    OWSSounds *instance = OWSSounds.sharedManager;
+    [instance.dbConnection setObject:@(sound) forKey:thread.uniqueId inCollection:kOWSSoundsStorageRingtoneCollection];
+}
+
+#pragma mark - Calls
+
++ (BOOL)shouldAudioPlayerLoopForSound:(OWSSound)sound
+{
+    return (sound == OWSSound_CallConnecting || sound == OWSSound_CallOutboundRinging ||
+        [self.allRingtoneSounds containsObject:@(sound)]);
+}
+
++ (nullable AVAudioPlayer *)audioPlayerForSound:(OWSSound)sound
+{
+    NSURL *soundURL = [OWSSounds soundURLForSound:sound];
+    NSError *error;
+    AVAudioPlayer *_Nullable player = [[AVAudioPlayer alloc] initWithContentsOfURL:soundURL error:&error];
```

Comment:
Same concerns as my earlier comment starting with: 

> Is this sufficient for background audio to resume when we're done?

Within the context of call audio, which this code may have been extracted from, we were playing fast and loose, because all call audio was book-ended by the CallAdaptee's starting/ending an audio activity. 

But now that we're using it outside of the context of a call, we'll need to properly terminate the audio session when done playing. Again - either by inlining startActivity/stopActivity calls somewhere in this class, or preferably, by using some kind of reusable OWSAudioPlayer class which accounts for this.

Reply:
+1 will do.